src/main/java/org/folio/services/consortia/ShadowInstanceSynchronizationHandler.java

```diff
@@ -0,0 +1,168 @@
+            .map(kafkaRecord.key())
+          : Future.succeededFuture(kafkaRecord.key()))
+        .onFailure(e -> LOG.warn(EVENT_HANDLING_ERROR_MSG, event.getTenant(), instanceId, e));
+    } catch (Exception e) {
+      LOG.warn("handle:: Error while handling event for shadow instances synchronization", e);
+      return Future.failedFuture(e);
+    }
+  }
+
+  private boolean isCentralTenantId(String tenantId, ConsortiumData consortiumData) {
+    return tenantId.equals(consortiumData.getCentralTenantId());
+  }
+
+  private Future<Void> synchronizeShadowInstances(DomainEvent<Instance> event, String instanceId,
+                                                  ConsortiumData consortiumData, Map<String, String> headers) {
+    return getShadowInstancesTenantIds(consortiumData.getConsortiumId(), consortiumData.getCentralTenantId(),
+      instanceId, headers)
+      .compose(tenantIds -> updateShadowInstances(event, tenantIds, headers));
+  }
+
+  private Future<List<String>> getShadowInstancesTenantIds(String consortiumId, String centralTenantId,
+                                                           String instanceId, Map<String, String> headers) {
+    String okapiUrl = headers.get(URL);
+    String preparedPath = format(SHARING_INSTANCES_PATH, consortiumId, instanceId);
+    WebClient client = WebClient.wrap(httpClient);
+    HttpRequest<Buffer> request = client.requestAbs(GET, okapiUrl + preparedPath);
+    headers.forEach(request::putHeader);
+    request.addQueryParam(LIMIT_QUERY_PARAM, TENANT_IDS_LIMIT);
+
+    return request.send().compose(response -> {
+      if (response.statusCode() != HTTP_OK) {
+        String msg = format("Error retrieving shadow instances tenant ids, response status: '%s', response body: '%s'",
+          response.statusCode(), response.bodyAsString());
+        LOG.warn("getShadowInstancesTenantIds:: {}", msg);
+        return Future.failedFuture(msg);
+      }
+
+      List<String> affiliationsTenantIds = response.bodyAsJsonObject().getJsonArray(SHARING_INSTANCES_FIELD)
+        .stream()
+        .map(JsonObject.class::cast)
+        .flatMap(sharing ->
+          Stream.of(sharing.getString(TARGET_TENANT_ID_FIELD), sharing.getString(SOURCE_TENANT_ID_FIELD)))
+        .filter(tenantId -> !tenantId.equals(centralTenantId))
+        .collect(Collectors.toList());
+      return Future.succeededFuture(affiliationsTenantIds);
+    });
+  }
+
+  private Future<Void> updateShadowInstances(DomainEvent<Instance> event, List<String> tenantIds,
+                                             Map<String, String> headers) {
+    LOG.info("updateShadowInstances:: Trying to update shadow instances in the following tenants: {} ", tenantIds);
+    ArrayList<Future<Void>> updateFutures = new ArrayList<>();
+    Instance instance = JsonObject.mapFrom(event.getNewEntity()).mapTo(Instance.class);
+    prepareInstanceForUpdate(instance);
+
+    for (String tenantId : tenantIds) {
```

Comment:
This has the potential to launch numerous call to update instances at the same time. Is it possible to constraint it?

Reply:
Maybe updating instances sequentially on a worker thread.

Reply:
ShadowInstanceSynchronizationVerticle is deployed as a worker verticle so it’s already executed using a thread from vertx worker thread pool (options.setWorker(true) deployment option is set in the InitApi class).

Reply:
I see. Makes it better that my original thought. There is still a gap in that the database connections can be exhausted when large manipulations occur with instances. Firing off 61 requests would mean potentially a maximum of 61 requests for a database connection. More than likely mod-inventory-storage is working while a data import job is running so the connection pool will be exhausted and might see error. Do you see my scenario?

Reply:
got it, I see. Was thinking about it. Each request for a shadow instance update uses connection from a separate connection pool of the particular member tenant. So each shadow instance update will use one connection from pool of particular member tenant and because of that pool for a particular member tenant should not be exhausted.

Reply:
For MOBIUS, we are planning to shared db pools. does that change anything?

Reply:
pls clarify does it mean one common shared pool for tenants or multiple pools (1 per tenant) that share connections between tenants?

Reply:
One common pool for all tenants in a module.

Reply:
got it, added changes that allow updating not more than 10 instances (by default) in parallel.
What max pool size is planned to be set for such pool?

Reply:
I don't know what the max pool size will be. Concrete values should be gotten from performance testing